### DISCUSSION TITLE
Version 5.0.4

### DISCUSSION THREAD

**Discussion by jamonholmgren** [mobxjs/mobx-state-tree#1818] [General]
This patch-level release fixes three issues with node reconciliation. Thanks to @adamkovalsky for the PR!

* Adds a `getReconcilationType` method to `BaseNode`, which returns the type which should be used for reconcilating the node. This is normally the node's `type`, but in the case of `snapshotProcessor` it is overridden to be the `snapshotProcessor` type. This is necessary so that the union type `reconcile` can correctly reconcile the "current" node first if it's possible. This fixes #1791.
* Fixes the `snapshotProcessor` type's `isValidSnapshot` and `is` methods to return a validation error and not to throw if the specified preprocessor fails, which can happen when checking if an arbitrary snapshot is valid input (e.g. when determining a union type). We can't expect preprocessors to accept any arbitrary input, so we must assume they can throw when provided with invalid snapshots. This also fixes #1777.
* Fixes the array's `areSame` utility to first check that the provided value is a valid snapshot before calling `isMatchingSnapshotId`, to avoid the preprocessor throwing an error if the provided snapshot is invalid.

If there are any problems with this release, please file an issue or leave comments in the attached Discussion! Thanks to everyone who contributed!

<hr /><em>This discussion was created from the release <a href='https://github.com/mobxjs/mobx-state-tree/releases/tag/v5.0.4'>Version 5.0.4</a>.</em>

**Comment by adamkovalsky** [comment]
@jamonholmgren thank you again for your maintenance of this project! Unfortunately soon after the release I've discovered a regression related to snapshot preprocessing, for which I've introduced PR #1822. Sadly the way snapshot preprocessors are implemented is quite flaky and makes these corner cases hard to detect.

I've tested this as thoroughly as I could, both on our production project and with tests. Hopefully a quick patch can be released to prevent regressions for any users of the package.

**Reply by bengry** [reply]
Just saw this issue after upgrading to 5.0.4 and using this wrapper on top of `types.snapshotProcessor`:
```ts
const ourTypes = {
  /**
   * Allows handling with unknown values.
   *
   * @param Type The original type
   * @param recoveryPlan a recovery plan for resolving unknown values
   */
  withRecoveryPlan<T>(
    Type: ISimpleType<T>,
    recoveryPlan: (badSnapshot: unknown) => T,
  ): ISimpleType<T extends undefined ? Exclude<T, undefined> : T> {
    return types.snapshotProcessor(Type, {
      preProcessor(snapshotValue: unknown) {
        if (Type.is(snapshotValue)) {
          return snapshotValue;
        }

        return recoveryPlan(snapshotValue);
      },
    }) as any;
  },
};
```

This resulted in an infinite call loop on every assignment to the property that used this type.

Putting the information here in case anyone else also faces this. Reverting back to MST 5.0.3 solved this.

**Reply by adamkovalsky** [reply]
This should also be fixed by PR #1822.

**Reply by bengry** [reply]
We're still seeing these issues in 5.0.5. Reverting back to 5.0.3 "fixes" it.

**Reply by adamkovalsky** [reply]
@bengry I've created a simple sandbox to try to reproduce your issue but it doesn't reproduce:

https://codesandbox.io/s/jovial-field-ud55u?file=/src/index.ts

Can you please update my attempt or create a new sandbox demonstrating the issue?

**Reply by bengry** [reply]
@adamkovalsky I've been able to reproduce it here: https://codesandbox.io/s/great-montalcini-5s2us
It can also be somewhat reduced to a more minimal reproduction, but I think this is good enough.

**Reply by adamkovalsky** [reply]
@bengry sorry for not replying sooner, I was hoping to find a solution for this but unfortunately I haven't been able to find the time to actually resolve this yet. Your sandbox actually had a lot of unrelated noise, the issue boils down to simply having a `union` wrapped in a `snapshotProcessor`. I've opened [issue #1849](https://github.com/mobxjs/mobx-state-tree/issues/1849) which also links to a straightforward sandbox reproduction, maybe you or someone else can have a go at fixing the issue.
